src/index/FTSAlgorithms.cpp

```diff
@@ -113,6 +113,84 @@ Index::WordEntityPostings FTSAlgorithms::intersect(
   return resultWep;
 }
 
+// _____________________________________________________________________________
+Index::WordEntityPostings FTSAlgorithms::crossIntersect(
+    const Index::WordEntityPostings& matchingContextsWep,
+    const Index::WordEntityPostings& eBlockWep) {
+  // Example:
+  // matchingContextsWep.wids_: 3 4 3 4 3
+  // matchingContextsWep.cids_: 1 4 5 5 7
+  // -----------------------------------
+  // eBlockWep.cids_          : 4 5 5 8
+  // eBlockWep.eids_          : 2 1 2 1
+  // ===================================
+  // resultWep.cids_          : 4 5 5 5 5
+  // resultWep.wids_          : 4 3 4 3 4
+  // resultWep.eids_          : 2 1 2 2 1
+  LOG(DEBUG)
+      << "Intersection to filter the word-entity postings from a block so that "
+      << "only entries remain where the entity matches. If there are multiple "
+      << "entries with the same eid, then the crossproduct of them remains.\n";
+  LOG(DEBUG) << "matchingContextsWep.cids_ size: "
+             << matchingContextsWep.cids_.size() << '\n';
+  LOG(DEBUG) << "eBlockWep.cids_ size: " << eBlockWep.cids_.size() << '\n';
+  Index::WordEntityPostings resultWep;
+  // Handle trivial empty case
+  if (matchingContextsWep.cids_.empty() || eBlockWep.cids_.empty()) {
+    return resultWep;
+  }
+  resultWep.wids_.reserve(eBlockWep.cids_.size());
+  resultWep.wids_.clear();
+  resultWep.cids_.reserve(eBlockWep.cids_.size());
+  resultWep.cids_.clear();
+  resultWep.eids_.reserve(eBlockWep.cids_.size());
+  resultWep.eids_.clear();
+  resultWep.scores_.reserve(eBlockWep.cids_.size());
+  resultWep.scores_.clear();
+
+  size_t i = 0;
+  size_t j = 0;
+
+  while (i < matchingContextsWep.cids_.size() && j < eBlockWep.cids_.size()) {
+    while (matchingContextsWep.cids_[i] < eBlockWep.cids_[j]) {
+      ++i;
+      if (i >= matchingContextsWep.cids_.size()) {
+        return resultWep;
+      }
+    }
+    while (eBlockWep.cids_[j] < matchingContextsWep.cids_[i]) {
+      ++j;
+      if (j >= eBlockWep.cids_.size()) {
+        return resultWep;
+      }
+    }
+    while (matchingContextsWep.cids_[i] == eBlockWep.cids_[j]) {
+      // Make sure we get all matching elements from the entity list (l2)
+      // that match the current context.
+      // If there are multiple elements for that context in l1,
+      // we can safely skip them unless we want to incorporate the scores
+      // later on.
+      size_t k = 0;
+      while (matchingContextsWep.cids_[i + k] == matchingContextsWep.cids_[i]) {
+        resultWep.wids_.push_back(matchingContextsWep.wids_[i + k]);
+        resultWep.cids_.push_back(eBlockWep.cids_[j]);
+        resultWep.eids_.push_back(eBlockWep.eids_[j]);
+        resultWep.scores_.push_back(eBlockWep.scores_[j]);
+        k++;
+        if (k >= matchingContextsWep.cids_.size()) {
+          break;
+        }
+      }
+      j++;
+      if (j >= eBlockWep.cids_.size()) {
+        break;
+      }
+    }
+    ++i;
+  }
+  return resultWep;
```

Comment:
This can be implementted using the (new) general
`ad_utility::zipperJoinWithUndef` in `JoinAlgorithms.h`. 
You basically only have to configure, what happens with the results. (There are no UNDEF values here, so many
of the arguments can be `noop`. I can also show you this in our meeting.